### PR TITLE
[moe] add quantile-balanced moe routing

### PR DESCRIPTION
**Pull request by shuhuayu**
## Summary

Implement quantile balancing for Kimi K3's MoE router as described in section 2.3.3 and Appendix D of the Kimi K3 technical report.

- Add a reusable `QuantileBalancedTopKRouter` and histogram estimator to the common MoE components. The router computes Top-(k+1) once, uses the first `k` experts for routing, and retains the final score as the cutoff for quantile estimation.
- Add a shared optimizer pre-hook that reduces per-expert required-bias histograms and solves for a mean-centered expert bias before each optimizer step.
- Wire the shared components into Kimi K3 and replace its sign-based load-balancing hook for all registered flavors.
- Add a two-router, two-rank GPU test covering router observation, distributed histogram reduction, bias estimation, and histogram reset.

## Design

For each token, the router obtains the biased Top-(k+1) scores. The first `k` expert IDs are used for dispatch, while the `(k+1)`-th score is the cutoff. Routing weights continue to come from the original unbiased router scores.

The balancer records `cutoff - score` for every token and expert in 1000 uniform bins over the Appendix D range `[min(expert_bias) - 1, max(expert_bias) + 1]`. At the optimizer pre-hook, layer histograms are stacked and summed with one collective per applicable process group. The `k / num_experts` quantile is interpolated within its crossing bin, and the resulting mean-centered bias overwrites `expert_bias_E`.

The reusable router and histogram state live with the common MoE components, and optimizer-step coordination lives with the other optimizer hooks. The histogram is non-persistent, so it follows module device movement without becoming checkpoint state. Kimi-specific configuration and ownership of the persistent expert-bias buffer remain in the Kimi model.

## Relation to #4412

This is an alternative implementation to #4412. It integrates cutoff collection into routing instead of attaching a forward hook, avoiding a second Top-k operation. It also uses the report's adaptive histogram range and 1000 bins, observes training forwards only, and exercises the actual distributed collective in a GPU test.

## Test plan

- `pytest tests/unit_tests/gpu/test_quantile_balance.py -q` (`1 passed`)
- `pytest tests/unit_tests/cpu/test_moe.py tests/unit_tests/cpu/test_optimizer_param_groups.py -q` (`23 passed`)
- Scoped pre-commit checks, including formatting, lint, and Pyrefly (`passed`)

Thanks @QIU023 for the earlier implementation and discussion in #4412.

### CODE REVIEW

**Comment by shuhuayu** [comment]
@QIU023 @JavaZeroo can you help review here?

File: torchtitan/models/kimi_k3/moe.py [hunk 1]
```diff
@@ -117,19 +117,36 @@ class Config(MoE.Config):
         routed_up: Linear.Config
 
     def __init__(self, config: Config):
+        if config.load_balance_coeff is not None:
+            raise ValueError(
+                "KimiLatentMoE cannot combine sign-based and quantile balancing."
+            )
         super().__init__(config)
+        del self.expert_bias_E
```
Comment (JavaZeroo) [inline torchtitan/models/kimi_k3/moe.py]: Maybe we should also `del tokens_per_expert_E`?

**Comment by Jeremalloch** [comment]
Super cool work! Excited to see this come to torchtitan!

Would it make sense to put the Quantile Balancing implementation under the generic MoE/router components rather than making it Kimi-specific? The QB algorithm is architecture-agnostic and used by some other models (Marin for now, but perhaps more in the future).

Megatron also keeps quantile balancing [architecture agnostic](https://github.com/NVIDIA/Megatron-LM/blob/b005bf14c46b62169533e755ca2d4e62fe6b7e0a/megatron/core/transformer/moe/router.py#L319).

The other point is that I could imagine this being useful for continued/mid-training where the data distribution changes substantially. Marin had an [experiment showing good expert stability](https://github.com/marin-community/marin/issues/6171) when changing the data distribution.

Maybe the Kimi wiring stays model-specific while the QB histogram/update/checkpoint machinery lives with the reusable MoE routing utilities?

**Comment by shuhuayu** [comment]
>Would it make sense to put the Quantile Balancing implementation under the generic MoE/router components rather than making it Kimi-specific? 

Make sense! I moved it to the shared modules, can you have a another look/review? @Jeremalloch 

>The other point is that I could imagine this being useful for continued/mid-training where the data distribution changes substantially. Marin had an https://github.com/marin-community/marin/issues/6171 when changing the data distribution.

Thanks for the pointer. Great to know quantile balancing is also helpful to deal with data distribution shift efficiently for mid-training.

File: torchtitan/models/common/moe.py [hunk 2]
```diff
@@ -386,6 +387,165 @@ def forward(
...
+            k=self.top_k + 1,
+            dim=-1,
+            sorted=True,
+        )
+        self.quantile_balancer.observe(
+            scores_TE,
+            topk_plus_one_scores[:, self.top_k :],
+            expert_bias_E,
+        )
+        return topk_plus_one_expert_ids[:, : self.top_k]
```
Comment (Jeremalloch) [inline torchtitan/models/common/moe.py]: Add a `contiguous()` call here. Otherwise, slicing the Top-(k+1) IDs produces noncontiguous indices.  This will cause errors when the dispatcher calls `.view(-1)` with any `k > 1`.

File: torchtitan/models/common/moe.py [hunk 3]
```diff
@@ -386,6 +387,165 @@ def forward(
...
+            ).to(torch.int64)
+            bin_indices_ET = bin_indices_TE.clamp_(0, self.num_bins - 1).transpose(0, 1)
+            self.required_bias_histogram_EB.scatter_add_(
+                1,
+                bin_indices_ET,
+                torch.ones_like(
+                    bin_indices_ET,
+                    dtype=self.required_bias_histogram_EB.dtype,
+                ),
+            )
```
Comment (Jeremalloch) [inline torchtitan/models/common/moe.py:507]: I hit `SpmdTypeError` on this because the histogram lacks mesh annotations, and the indices carry the DP axis.

This either needs an explicit SPMD layout or a local observation boundary.

File: torchtitan/models/common/moe.py (hunk 2 above)
Comment (shuhuayu) [reply torchtitan/models/common/moe.py]: you are right, the test used `k=1` and missed this.

File: torchtitan/models/common/moe.py (hunk 3 above)
Comment (shuhuayu) [reply torchtitan/models/common/moe.py:507]: explicit annotation here is kind of tricky, `bin_indices_ET` inherits a `V` type from `scores_TE`, and there is no upstream rule for `scatter_add_`. i would set a local boundary here rather than a spmd type check.

**Review by Jeremalloch — APPROVED** [review]
LGTM.  I'm looking forward to trying this out for midtraining purposes.

I appreciate you taking the feedback into consideration to generalize it beyond Kimi K3.

File: torchtitan/models/common/moe.py [hunk 4]
```diff
@@ -318,6 +320,161 @@ def forward(
         )
 
 
+class QuantileBalancedTopKRouter(TokenChoiceTopKRouter):
+    """Top-k router that uses a biased Top-(k+1) cutoff during training."""
+
+    @dataclass(kw_only=True, slots=True)
+    class Config(TokenChoiceTopKRouter.Config):
+        num_bins: int = 1000
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: how did you choose this default?

Kimi K3 paper says
> Because counts are additive, the histogram represents the pooled global batch regardless of how tokens are sharded, so the estimate reflects the whole-batch quantile up to the bin width, at a communication cost of only **a few hundred bins per expert**.

File: torchtitan/models/common/moe.py [hunk 5]
```diff
@@ -318,6 +320,161 @@ def forward(
         )
 
 
+class QuantileBalancedTopKRouter(TokenChoiceTopKRouter):
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py:363]: Maybe we should add it in Kimi K3 folder first. DeepSeek v3 / v4 have their own routers as well.

File: torchtitan/models/common/moe.py [hunk 6]
```diff
@@ -318,6 +320,161 @@ def forward(
...
+    For scores bounded between zero and one, required expert biases lie between
+    the current minimum bias minus one and maximum bias plus one. Each training
+    micro-batch is accumulated into uniform bins over that interval.
+    """
+
+    @dataclass(kw_only=True, slots=True)
+    class Config(Module.Config):
+        num_experts: int
+        top_k: int
+        num_bins: int = 1000
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: shouldn't have default, as router already has

**Review by tianyu-l — COMMENTED** [review]
code looks quite reasonable; didn't check math

File: torchtitan/models/common/moe.py (hunk 5 above)
Comment (tianyu-l) [reply torchtitan/models/common/moe.py:363]: I saw
> I appreciate you taking the feedback into consideration to generalize it beyond Kimi K3.

But I think even if we move it under K3, it's usable from outside.

File: torchtitan/models/common/moe.py (hunk 4 above)
Comment (shuhuayu) [reply torchtitan/models/common/moe.py]: The `B=1000` shows in the Section D Properties, but i agree we should remove the defaults as router does.

File: torchtitan/models/common/moe.py (hunk 5 above)
Comment (shuhuayu) [reply torchtitan/models/common/moe.py:363]: I would prefer to put quantile balancing module in the shared place. deepseek v3/4 defines model specific expert selection logic locally, but reuses the common `register_moe_load_balancing_hook`. If we want to make everything local, we should also move the `register_moe_quantile_balancing_hook` to from `optimizer.py` to k3 folder, so the current code organization would be cleaner. 

but i do get your point if a feature is used only by one model it should stay in model's local folder, my point is more from the newly added hook, which seems to be cleaner to me if stay in the common `optimizer.py`. wdyt?

**Comment by shuhuayu** [comment]
the changes in `test_moe.py` is a simple two-line fix to a recent pr https://github.com/pytorch/torchtitan/pull/4594

File: torchtitan/models/common/moe.py (hunk 5 above)
Comment (tianyu-l) [reply torchtitan/models/common/moe.py:363]: In fact, I'm planning a refactor to probably move the register hook functions away from optimizer.py, because I don't think it's optimizer responsibility anyway. That said, I'm OK with this PR being merged as is.

File: torchtitan/models/common/moe.py [hunk 7]
```diff
@@ -352,6 +361,168 @@ def forward(
...
+        self.num_bins = config.num_bins
+        self.register_buffer(
+            "required_bias_histogram_EB",
+            torch.zeros(config.num_experts, config.num_bins, dtype=torch.int32),
+            persistent=False,
+        )
+
+    @staticmethod
+    def _local_tensor(tensor: torch.Tensor) -> torch.Tensor:
+        return tensor.to_local() if isinstance(tensor, DTensor) else tensor
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: DTensor in fwd-bwd has been deprecated. Could we remove?

File: torchtitan/models/common/moe.py [hunk 8]
```diff
@@ -352,6 +361,168 @@ def forward(
...
+
+    def observe(
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py:443]: is this being tested in CI integrations tests with spmd_typechecking on?

**Review by tianyu-l — APPROVED** [review]
please address comments before merging

File: torchtitan/models/common/moe.py (hunk 7 above)
Comment (shuhuayu) [reply torchtitan/models/common/moe.py]: that's cool. removed the dtensor convertor.

File: torchtitan/models/common/moe.py (hunk 8 above)
Comment (shuhuayu) [reply torchtitan/models/common/moe.py:443]: yes.
